### PR TITLE
Refactoring TableStorage and BlobContainer

### PR DESCRIPTION
**Pull request by kapil-ms**

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@kapil-ms__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (kaagarw). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: Common/Helpers/BlobStorageClient.cs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+namespace Microsoft.Azure.Devices.Applications.RemoteMonitoring.Common.Helpers
+{
+    /// <summary>
+    /// Helper methods, related to blob storage.
+    /// </summary>
+    public class BlobStorageClient : IBlobStorageClient
+    {
+        private CloudStorageAccount _storageAccount;
+        private CloudBlobContainer _container;
+        private CloudBlockBlob _blob;
```
Comment (rraomsft) [inline Common/Helpers/BlobStorageClient.cs]: CloudBlockBlob should not be stored as state in the client.

File: Common/Helpers/BlobStorageClient.cs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+            _storageAccount = CloudStorageAccount.Parse(connectionString);
+            _blobClient = _storageAccount.CreateCloudBlobClient();
+            _containerName = containerName;
+            _blobName = blobName;
+        }
+        
+        public async Task UploadFromByteArrayAsync(byte[] buffer, int index, int count, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext)
+        {
+            await this.CreateCloudBlockBlobAsync();
+            await _blob.UploadFromByteArrayAsync(
```
Comment (rraomsft) [inline Common/Helpers/BlobStorageClient.cs]: CreateCloudBlockBlobAsync should not update state of the class. Instead return the blob and UploadFromByteArrayAsync should use the blob instance directly.

File: Common/Helpers/AzureTableStorageClient.cs [hunk 3]
```diff
@@ -10,37 +11,63 @@
...
         private CloudTable _table;
+        private readonly CloudTableClient _tableClient;
+        private readonly string _tableName;
 
-        public AzureTableStorageHelper(string storageConnectionString, string tableName)
+        public AzureTableStorageClient(string storageConnectionString, string tableName)
         {
-            _storageConnectionString = storageConnectionString;
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(storageConnectionString);
+            _tableClient = storageAccount.CreateCloudTableClient();
```
Comment (handeadiguzel) [inline Common/Helpers/AzureTableStorageClient.cs]: let's start using "this._tableClient" to differentiate between local variables and class variables.

Comment (handeadiguzel) [inline Common/Helpers/AzureTableStorageClient.cs]: actually if you use resharper's code cleanup, it adds "this." for you

File: Common/Helpers/AzureTableStorageClient.cs [hunk 4]
```diff
@@ -10,37 +11,63 @@
...
+            if (_table == null && _tableName != null)
+            {
+                _table = _tableClient.GetTableReference(_tableName);
+                await _table.CreateIfNotExistsAsync();
+            }
             return _table;
         }
-
-        public CloudTable GetTable()
+        private CloudTable GetCloudTable()
```
Comment (handeadiguzel) [inline Common/Helpers/AzureTableStorageClient.cs]: are both of these used: GetCloudTable and GetCloudTable? 
Calling: GetCloudTableAsync().Wait() would wait synchronously on the async create task and you wouldn't need the GetCloudTable() method

File: Common/Helpers/AzureTableStorageClient.cs [hunk 5]
```diff
@@ -10,37 +11,63 @@
...
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(storageConnectionString);
+            _tableClient = storageAccount.CreateCloudTableClient();
             _tableName = tableName;
-            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(_storageConnectionString);
-            CloudTableClient tableClient = storageAccount.CreateCloudTableClient();
-            _table = tableClient.GetTableReference(_tableName);
         }
-
-        public async Task<CloudTable> GetTableAsync()
+        private async Task<CloudTable> GetCloudTableAsync()
```
Comment (handeadiguzel) [inline Common/Helpers/AzureTableStorageClient.cs]: I usually place private functions after public ones.

File: Common/Helpers/AzureTableStorageClient.cs [hunk 6]
```diff
@@ -10,37 +11,63 @@
...
+        {
+            CloudTable table = this.GetCloudTable();
+            return table.Execute(tableOperation);
+        }
+        public async Task<TableResult> ExecuteAsync(TableOperation operation)
+        {
+            CloudTable table = await this.GetCloudTableAsync();
+            return await table.ExecuteAsync(operation);
+        }
+        public IEnumerable<T> ExecuteQuery<T>(TableQuery<T> tableQuery) where T : TableEntity, new()
```
Comment (handeadiguzel) [inline Common/Helpers/AzureTableStorageClient.cs]: where T : ITableEntity ?

File: Common/Helpers/AzureTableStorageClient.cs [hunk 7]
```diff
@@ -10,37 +11,63 @@
...
-            _table = tableClient.GetTableReference(_tableName);
         }
-
-        public async Task<CloudTable> GetTableAsync()
+        private async Task<CloudTable> GetCloudTableAsync()
         {
-            await _table.CreateIfNotExistsAsync();
+            if (_table == null && _tableName != null)
+            {
+                _table = _tableClient.GetTableReference(_tableName);
```
Comment (handeadiguzel) [inline Common/Helpers/AzureTableStorageClient.cs]: if you know the tableName in the constructor why not initialize tableClient there? and the methods will access the table as: this._table

File: Common/Helpers/BlobStorageClient.cs [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        {
+            _storageAccount = CloudStorageAccount.Parse(connectionString);
+            _blobClient = _storageAccount.CreateCloudBlobClient();
+            _containerName = containerName;
+            _blobName = blobName;
+        }
+        
+        public async Task UploadFromByteArrayAsync(byte[] buffer, int index, int count, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext)
+        {
+            await this.CreateCloudBlockBlobAsync();
```
Comment (handeadiguzel) [inline Common/Helpers/BlobStorageClient.cs]: so all the methods need to know that they have to call CreateCloudBlockBlobAsync method first... I would instead create the blob in the constructor.

File: Common/Helpers/BlobStorageClient.cs [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+                    await _blob.DownloadToByteArrayAsync(existingBytes, 0);
+                    return existingBytes;
+                }
+            }
+            return null;
+        }
+
+        public async Task<string> GetBlobEtag()
+        {
+            await this.CreateCloudBlockBlobAsync();
```
Comment (handeadiguzel) [inline Common/Helpers/BlobStorageClient.cs]: I think this is a little confusing. For someone that calls GetBlobEtag method they assume that the blob is already created and they want to access its etag. Same comment but I think this should be a 1 time operation: CreateCloudBlockBlobAsync

File: Simulator/Simulator.WebJob/SimulatorModule.cs [hunk 10]
```diff
@@ -63,11 +63,11 @@ protected override void Load(ContainerBuilder builder)
...
 
-            builder.RegisterType<BlobStorageHelper>()
-                .As<IBlobStorageHelper>();
+            builder.RegisterType<AzureTableStorageClientFactory>()
+                .As<IAzureTableStorageClientFactory>();
 
-            builder.RegisterType<AzureTableStorageHelper>()
-                .As<IAzureTableStorageHelper>();
+            builder.RegisterType<BlobStorageClientFactory>()
+                .As<IBlobStorageClientFactory>();
```
Comment (handeadiguzel) [inline Simulator/Simulator.WebJob/SimulatorModule.cs]: only simulator module needs to be updated?

File: Common/Common.csproj [hunk 11]
```diff
@@ -164,10 +164,16 @@
     <Compile Include="Extensions\StringExtensions.cs" />
     <Compile Include="Factory\SampleDeviceFactory.cs" />
     <Compile Include="Helpers\AzureRetryHelper.cs" />
-    <Compile Include="Helpers\AzureTableStorageHelper.cs" />
-    <Compile Include="Helpers\BlobStorageHelper.cs" />
-    <Compile Include="Helpers\IAzureTableStorageHelper.cs" />
-    <Compile Include="Helpers\IBlobStorageHelper.cs" />
+    <Compile Include="Helpers\AzureTableStorageClient.cs" />
```
Comment (handeadiguzel) [inline Common/Common.csproj]: make sure to run unit tests and RM and do some manual smoke testing...

File: Common/Helpers/BlobStorageClient.cs (hunk 1 above)
Comment (kapil-ms) [inline Common/Helpers/BlobStorageClient.cs]: done

File: Common/Helpers/BlobStorageClient.cs (hunk 2 above)
Comment (kapil-ms) [inline Common/Helpers/BlobStorageClient.cs]: done

File: Common/Helpers/AzureTableStorageClient.cs (hunk 3 above)
Comment (kapil-ms) [inline Common/Helpers/AzureTableStorageClient.cs]: ran cleanup, however it didn't add this keyword.

File: Common/Helpers/AzureTableStorageClient.cs (hunk 5 above)
Comment (kapil-ms) [inline Common/Helpers/AzureTableStorageClient.cs]: cleanup fixed it

File: Common/Helpers/AzureTableStorageClient.cs (hunk 6 above)
Comment (kapil-ms) [inline Common/Helpers/AzureTableStorageClient.cs]: T must be a non-abstract type with a public parameterless constructor, so cannot use ITableEntity

File: Common/Helpers/AzureTableStorageClient.cs (hunk 7 above)
Comment (kapil-ms) [inline Common/Helpers/AzureTableStorageClient.cs]: so some functions use async version and some don't, having separate create and createasync methods allows to not change this behavior.

File: Common/Helpers/BlobStorageClient.cs (hunk 8 above)
Comment (kapil-ms) [inline Common/Helpers/BlobStorageClient.cs]: as discussed with @rraomsft , BlobStorageClient will not maintain blob state

File: Common/Helpers/BlobStorageClient.cs (hunk 9 above)
Comment (kapil-ms) [inline Common/Helpers/BlobStorageClient.cs]: same as above, now taking blob name, and then creating the blob for lucidity.

File: Simulator/Simulator.WebJob/SimulatorModule.cs (hunk 10 above)
Comment (kapil-ms) [inline Simulator/Simulator.WebJob/SimulatorModule.cs]: all 3 should be updated

File: Common/Common.csproj (hunk 11 above)
Comment (kapil-ms) [inline Common/Common.csproj]: did unit testing, ran RM, and postman
all working

**Pull request merged**